Zend/zend_timer.c

```diff
@@ -0,0 +1,116 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Kévin Dunglas <kevin@dunglas.dev>                            |
+   +----------------------------------------------------------------------+
+ */
+
+#ifdef ZEND_TIMERS
+
+#include <stdio.h>
+#include <signal.h>
+#include <time.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+
+#include "zend.h"
+#include "zend_globals.h"
+
+// Musl Libc defines this macro, glibc does not
+// According to "man 2 timer_create" this field should always be available, but it's not: https://sourceware.org/bugzilla/show_bug.cgi?id=27417
+# ifndef sigev_notify_thread_id
+# define sigev_notify_thread_id _sigev_un._tid
+# endif
+
+ZEND_API void zend_timer_init(void) /* {{{ */
+{
+	struct sigevent sev;
+	sev.sigev_notify = SIGEV_THREAD_ID;
+	sev.sigev_value.sival_ptr = &EG(timer);
+	// The chosen signal must:
+	// 1. not be used internally by libc
+	// 2. be allowed to happen spuriously without consequences
+	// 3. not be commonly used by applications, this excludes SIGALRM, SIGUSR1 and SIGUSR2
+	// 4. not be used by profilers, this excludes SIGPROF
+	// 5. not be used internally by runtimes of programs that can embed PHP, this excludes SIGURG, which is used by Go
```

Comment:
Do the realtime signals work here? It's up to the application to decide what to do with them, which seems better than a standard signal. They work a little differently from standard signals, though:
https://man7.org/linux/man-pages/man7/signal.7.html